ensure MatchNodeSelectorTerms() runs statelessly

Huang-Wei opened the pull request:
**What this PR does**:

Fix sorting behavior in selector.go:

- move sorting from NewRequirement() out to String()
- add related unit tests
- add unit tests in one of outer callers (pkg/apis/core/v1/helper)

**Why we need it**:
- Without this fix, scheduling and daemonset controller doesn't work well in some (corner) cases

**Which issue(s) this PR fixes**:
Fixes #66298

**Special notes for your reviewer**:
Parameter `nodeSelectorTerms` in method MatchNodeSelectorTerms() is a slice, which is fundamentally a {*elements, len, cap} tuple - i.e. it's passing in a pointer. In that method, NodeSelectorRequirementsAsSelector() -> NewRequirement() is invoked, and the `matchExpressions[*].values` is passed into and **modified** via `sort.Strings(vals)`.

This will cause following daemonset pod fall into an infinite create/delete loop:

```yaml
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: problem
spec:
  selector:
    matchLabels:
      app: sleeper
  template:
    metadata:
      labels:
        app: sleeper
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: kubernetes.io/hostname
                operator: In
                values:
                - 127.0.0.2
                - 127.0.0.1
      containers:
      - name: busybox
        image: busybox
        command: ["/bin/sleep", "7200"]
```

(the problem can be stably reproduced on a local cluster started by `hack/local-up-cluster.sh`)

The first time daemonset yaml is handled by apiserver and persisted in etcd with original format (original order of values was kept - 127.0.0.2, 127.0.0.1). After that, daemonset controller tries to schedule pod, and it reuses the predicates logic in scheduler component - where the values are **sorted** deeply. This not only causes the pod to be created in sorted order (127.0.0.1, 127.0.0.2), but also introduced a bug when updating daemonset - internally ds controller use a "rawMessage" (bytes of an object) to calculate hash acting as a "controller-revision-hash" to control revision rollingUpdate/rollBack, so it keeps killing "old" pod and spawning "new" pod back and forth, and fall into an infinite loop.

The issue exists in `master`, `release-1.11` and `release-1.10`.

**Release note**:
```release-note
NONE
```

Huang-Wei committed:
fix sorting behavior in selector.go

- move sorting from NewRequirement() out to String()
- add related unit tests
- add unit tests in one of outer callers (pkg/apis/core/v1/helper)

Closes #66298

Huang-Wei commented:
/sig scheduling
/sig apps

Huang-Wei commented:
/cc @k82cn @janetkuo

yastij commented:
/ok-to-test

Huang-Wei commented:
/test pull-kubernetes-e2e-kops-aws

Huang-Wei commented:
/retest

k82cn commented on pkg/apis/core/v1/helper/helpers.go:
```diff
@@ -304,7 +304,11 @@ func MatchNodeSelectorTerms(
+		// if we don't make a copy
+		// in NodeSelectorRequirementsAsSelector(), NewRequirement() will
+		// modify input values via sort.Strings(vals)
+		req := r.DeepCopy()
```
Good catch ! 

IMO, it's better to enhance `NodeSelectorRequirementsAsSelector` to avoid touching the parameters. @liggitt , WDYT?

liggitt replied:
Agree, good catch. I do think we want to avoid a full deep copy like this if we can… this is evaluated frequently and could be a significant allocation hotspot. Can we remove the need for mutation or just the string slice before sort?

Huang-Wei replied:
@k82cn maybe we can walk one step further - make changes in `NewRequirement()` (in selector.go) as there are other invocations like TopologySelectorRequirementsAsSelector().

 @liggitt make a copy of string slice before sort (`NewRequirement()` in selector.go) might be a safe solution. I'm not confident at removing the whole sort thing... (suppose this is what you mean by "remove the need for mutation")

yastij replied:
Good catch !
I’d be more in favor of just removing the sort.

k82cn replied:
>  just the string slice before sort?

Both value & key are sorted, any background for that?  There's a comment that `// sort to have deterministic string representation`, do we need to keep that?



https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/labels/selector.go#L851

Huang-Wei commented:
Appreciate everyone's comment. Probably we need another PR to address the issue in apimachinery code, so let's continue the discussion under the PR, rather than the inline code.

@deads2k would you kindly comment on this issue? The background is that, for a unordered `nodeSelectorTerms` (given by user in node affinity of a Daemonset yaml let's say), internally it calls selector.go (in apimachinery/pkg/labels). As the values is a slice, internally sorting behavior in selector.go actually mutates the values in the input Daemonset/Deployment spec - which causes some interesting bug like #66298.

Basically we have several options to solve this problem:

1. one is to make a copy outside (proposed in this PR), so that after calling selector.go, its values aren't modified
1. the other one (commented above by other guys), is to get clear about the real usecase/motivation of sorting in selector.go. If it's not that necessary, maybe we can:
    1. simply remove related sorting logic
    1. or more safely, make a slice copy before sorting

cc @liggitt as well

deads2k commented:
> the other one (commented above by other guys), is to get clear about the real usecase/motivation of sorting in selector.go. If it's not that necessary, maybe we can:
> simply remove related sorting logic

I found where it was introduced https://github.com/kubernetes/kubernetes/pull/35194 and I followed up by checking the 17 usages.  I don't see a reason why it is sorted.  I opened https://github.com/kubernetes/kubernetes/pull/66580

yastij commented:
@Huang-Wei - can you update your PR and remove the sort ? Thanks !

Huang-Wei commented:
@yastij the code has been updated according to discussion in https://github.com/kubernetes/kubernetes/pull/66580#issuecomment-407793217.

Fix sorting behavior in selector.go:

- move sorting from NewRequirement() out to String()
- add related unit tests
- add unit tests in one of outer callers (pkg/apis/core/v1/helper)

Any comment is welcome.

/assign @liggitt @deads2k

yastij commented on staging/src/k8s.io/apimachinery/pkg/labels/selector.go:
```diff
@@ -166,7 +166,7 @@ func NewRequirement(key string, op selection.Operator, vals []string) (*Requirem
 		}
 	}
-	sort.Strings(vals)
+	// sort.Strings(vals) // move sorting behavior out
```
You can delete this line.

Huang-Wei replied:
got it. done.

yastij reviewed (commented):
Just a nit. Otherwise lgtm

Huang-Wei commented:
Above bazel test failure is because we moved sorting() to `String()` of selector, so `reflect.DeepEqual()` doesn't invoke `String()`. And unsurprisingly, in log message they look the same - that's because `t.Errorf()` invokes `String()`:

```
I0725 17:39:01.388] --- FAIL: TestLabelSelectorAsSelector (0.00s)
I0725 17:39:01.388] 	helpers_test.go:81: [3]expected:
I0725 17:39:01.388] 			baz in (norf,qux)
I0725 17:39:01.389] 		but got:
I0725 17:39:01.389] 			baz in (norf,qux)
I0725 17:39:01.389] 	helpers_test.go:81: [4]expected:
I0725 17:39:01.389] 			baz in (norf,qux),foo=bar
I0725 17:39:01.389] 		but got:
I0725 17:39:01.390] 			baz in (norf,qux),foo=bar
```

changed to call String() explicitly. (or, change the "expected" values to be same order as input)

```diff
@@ -77,7 +78,7 @@ func TestLabelSelectorAsSelector(t *testing.T) {
                if err != nil && !tc.expectErr {
                        t.Errorf("[%v]did not expect error but got: %v", i, err)
                }
-               if !reflect.DeepEqual(out, tc.out) {
+               if !reflect.DeepEqual(fmt.Sprint(out), fmt.Sprint(tc.out)) {
                        t.Errorf("[%v]expected:\n\t%+v\nbut got:\n\t%+v", i, tc.out, out)
                }
        }
```

Huang-Wei commented:
/retest

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers_test.go:
```diff
@@ -77,7 +78,7 @@ func TestLabelSelectorAsSelector(t *testing.T) {
 		if err != nil && !tc.expectErr {
 			t.Errorf("[%v]did not expect error but got: %v", i, err)
 		}
-		if !reflect.DeepEqual(out, tc.out) {
```
this doesn't look right... what in LabelSelectorAsSelector was not preserving the existing data?

Huang-Wei replied:
LabelSelectorAsSelector preserves the existing data, it's just the testcase that expect the input data being ordered.

For an input like `"qux", "norf"`:
https://github.com/kubernetes/kubernetes/blob/9d4b574699dfe198d930a6c3282d188581b59f8b/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers_test.go#L29-L33

It's expected to be `"baz in (norf,qux),foo=bar"`:

https://github.com/kubernetes/kubernetes/blob/9d4b574699dfe198d930a6c3282d188581b59f8b/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers_test.go#L56-L59

So as I commented in https://github.com/kubernetes/kubernetes/pull/66480#issuecomment-407844550, it makes sense to change the testcase expected ouput?

> changed to call String() explicitly. (**or, change the "expected" values to be same order as input**)

liggitt replied:
I see... we could do both...
* have one test case that changes the expected output: `out: mustParse("baz in (qux,norf),foo=bar")`
* have one test case that makes sure `out.String()` and `tc.out.String()` match, regardless of the order of `tc.out`

Huang-Wei replied:
How about this: keep input data as is, and for each test, separate semantics check and string check:

- for input, add semantics check to ensure it's not modifed
- for output, compare to expectedOutput using string check

```diff
        for i, tc := range tc {
+               inCopy := tc.in.DeepCopy()
                out, err := LabelSelectorAsSelector(tc.in)
+               // after calling LabelSelectorAsSelector, tc.in shouldn't be modified
+               if !reflect.DeepEqual(inCopy, tc.in) {
+                       t.Errorf("[%v]expected:\n\t%#v\nbut got:\n\t%#v", i, inCopy, tc.in)
+               }
                if err == nil && tc.expectErr {
                        t.Errorf("[%v]expected error but got none.", i)
                }
                if err != nil && !tc.expectErr {
                        t.Errorf("[%v]did not expect error but got: %v", i, err)
                }
-               if !reflect.DeepEqual(fmt.Sprint(out), fmt.Sprint(tc.out)) {
-                       t.Errorf("[%v]expected:\n\t%+v\nbut got:\n\t%+v", i, tc.out, out)
+               // fmt.Sprint() over String() as nil.String() will panic
+               if fmt.Sprint(out) != fmt.Sprint(tc.out) {
+                       t.Errorf("[%v]expected:\n\t%s\nbut got:\n\t%s", i, fmt.Sprint(tc.out), fmt.Sprint(out))
                }
        }
 }
```

Huang-Wei commented:
/retest

k82cn commented on staging/src/k8s.io/apimachinery/pkg/labels/selector.go:
```diff
@@ -166,7 +166,6 @@ func NewRequirement(key string, op selection.Operator, vals []string) (*Requirem
 			return nil, err
 		}
 	}
-	sort.Strings(vals)
```
I'm not sure which option is better: if sort in `NewRequirement`, it only sort once; if sort in `String()`, it'll sort it every time when call `String()`.

According to the description at #35194, it improve performance as follow:

> Early results:
>
> * scheduler throughput it ~15% higher
> * apiserver cpu-usage is also lower (seems to be also ~10-15%)

IMO, avoiding sort in `String()` by removing `set.Strings.List()` should contribute :)

k82cn replied:
Would you try benchmark of scheduler? I'm not sure apiserver, but I think there should be some tools there.

Huang-Wei replied:
I tried benchmark test of scheduler, but didn't find any significant difference between the two solutions. And I realized that it might be current benchmark testcases doesn't cover the scenario we try to fix - multiple-value requirements. So I added a case like this: ([]string{"yellow", "green"} are designed in particular to hit sort())

```diff
diff --git a/test/integration/scheduler_perf/scheduler_bench_test.go b/test/integration/scheduler_perf/scheduler_bench_test.go
index fb28da4ec6..6ab25056d2 100644
--- a/test/integration/scheduler_perf/scheduler_bench_test.go
+++ b/test/integration/scheduler_perf/scheduler_bench_test.go
@@ -71,7 +71,27 @@ func BenchmarkSchedulingAntiAffinity(b *testing.B) {
 			benchmarkScheduling(test.nodes, test.existingPods, test.minPods, setupStrategy, testStrategy, b)
 		})
 	}
+}
 
+// BenchmarkSchedulingNodeAffinity benchmarks the scheduling rate of pods with
+// NodeAffinity rules when the cluster has various quantities of nodes and
+// scheduled pods.
+func BenchmarkSchedulingNodeAffinity(b *testing.B) {
+	tests := []struct{ nodes, existingPods, minPods int }{
+		{nodes: 500, existingPods: 250, minPods: 250},
+		{nodes: 500, existingPods: 5000, minPods: 250},
+	}
+	// The setup strategy creates pods with no affinity rules.
+	setupStrategy := testutils.NewSimpleWithControllerCreatePodStrategy("setup")
+	// The test strategy creates pods with node-affinity for each other.
+	testBasePod := makeBasePodWithNodeAffinity("color", []string{"yellow", "green"})
+	testStrategy := testutils.NewCustomCreatePodStrategy(testBasePod)
+	for _, test := range tests {
+		name := fmt.Sprintf("%vNodes/%vPods", test.nodes, test.existingPods)
+		b.Run(name, func(b *testing.B) {
+			benchmarkScheduling(test.nodes, test.existingPods, test.minPods, setupStrategy, testStrategy, b)
+		})
+	}
 }
 
 // makeBasePodWithAntiAffinity creates a Pod object to be used as a template.
@@ -99,6 +119,35 @@ func makeBasePodWithAntiAffinity(podLabels, affinityLabels map[string]string) *v
 	return basePod
 }
 
+// makeBasePodWithNodeAffinity creates a Pod object to be used as a template.
+// The Pod has a NodeAffinity requirement against nodes with the given expressions.
+func makeBasePodWithNodeAffinity(key string, vals []string) *v1.Pod {
+	basePod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "node-affinity-",
+		},
+		Spec: testutils.MakePodSpec(),
+	}
+	basePod.Spec.Affinity = &v1.Affinity{
+		NodeAffinity: &v1.NodeAffinity{
+			RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+				NodeSelectorTerms: []v1.NodeSelectorTerm{
+					{
+						MatchExpressions: []v1.NodeSelectorRequirement{
+							{
+								Key:      key,
+								Operator: v1.NodeSelectorOpNotIn,
+								Values:   vals,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	return basePod
+}
+
 // benchmarkScheduling benchmarks scheduling rate with specific number of nodes
 // and specific number of pods already scheduled.
 // This will schedule numExistingPods pods before the benchmark starts, and at
```

Then I rerun it (the command is `RUN_BENCHMARK=true ./test-performance.sh`) a few times, still no big difference on the benchmark test result. They're both sth. like:

```
BenchmarkScheduling/100Nodes/0Pods-8                 200           7131056 ns/op
BenchmarkScheduling/100Nodes/1000Pods-8              100          12868582 ns/op
BenchmarkScheduling/1000Nodes/0Pods-8                100          56496372 ns/op
BenchmarkScheduling/1000Nodes/1000Pods-8             100          52169635 ns/op
BenchmarkSchedulingAntiAffinity/500Nodes/250Pods-8                   250          16827985 ns/op
BenchmarkSchedulingAntiAffinity/500Nodes/5000Pods-8                  250          25059125 ns/op
BenchmarkSchedulingNodeAffinity/500Nodes/250Pods-8                   250          13460290 ns/op
BenchmarkSchedulingNodeAffinity/500Nodes/5000Pods-8                  250          21225831 ns/op

0s      rate: 621       total: 621 (qps frequency: map[621:1])
1s      rate: 753       total: 1374 (qps frequency: map[621:1 753:1])
2s      rate: 776       total: 2150 (qps frequency: map[753:1 776:1 621:1])
3s      rate: 790       total: 2940 (qps frequency: map[753:1 776:1 790:1 621:1])
Scheduled 3000 Pods in 4 seconds (750 per second on average). min QPS was 621
```

Then I did another test to print a message in both solutions (right before `sort.Strings()`). The result shows that:

- for the solution to put safeSort() in NewReuirement, the message shows up `30993 ` times
- for the solution to put safeSort() in String(), the message shows up `0` times

So it shows although the first solution runs sorting for 30k times, the benchmark didn't show major differences - I guess it's b/c we're justing sorting a two elements slice, that's not a big deal - but it's a common case, I believe nobody would use an affinity having a big values set as "matchExpressions".

k82cn replied:
if so, both are ok to me :)

seh replied:
> I believe nobody would use an affinity having a big values set as "matchExpressions".

I'm not sure what you think is big, but in the motivating #66298, [my case has nine elements today](https://github.com/kubernetes/kubernetes/issues/66298#issuecomment-405651546), and will likely accumulate more in the near future.

Huang-Wei replied:
@seh I can't tell an exact number without running benchmark tests. But sorting dozens of elements would be trivial. And also note that in current solution (safeSort() in String() of Requirement), it doesn't hit String() - internal objects are kept what it is - so there is no sorting.

Huang-Wei commented:
@liggitt @deads2k do you have any other comments?

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/labels/selector.go:
```diff
@@ -299,7 +298,8 @@ func (r *Requirement) String() string {
 		buffer.WriteString(r.strValues[0])
 	} else { // only > 1 since == 0 prohibited by NewRequirement
-		buffer.WriteString(strings.Join(r.strValues, ","))
+		// move sorting behavior from NewRequirement() to here
```
Please make the comment describe why this is the right place, not where you moved it from.

Huang-Wei replied:
done.

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/labels/selector_test.go:
```diff
@@ -573,3 +573,39 @@ func TestAdd(t *testing.T) {
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if got := safeSort(tt.in); !reflect.DeepEqual(got, tt.want) {
```
This test doesn't verify that tt.in is unmodified?

Huang-Wei replied:
good point. addressed.

Huang-Wei commented:
@lavalamp Thanks for the comments. All comments have been addressed. PTAL.

liggitt commented:
/lgtm
